File: backend/app/models/flashcard.py
from sqlalchemy import Column, Float, Integer, String, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship

from app.database import Base


class FlashcardDeck(Base):
    __tablename__ = "flashcard_decks"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    course_id = Column(Integer, ForeignKey("courses.id"), nullable=True)
    created_at = Column(DateTime, server_default=func.now())

    cards = relationship(
        "Flashcard",
        back_populates="deck",
        cascade="all, delete-orphan",
        order_by="Flashcard.id",
    )


class Flashcard(Base):
    __tablename__ = "flashcards"

    id = Column(Integer, primary_key=True, index=True)
    deck_id = Column(Integer, ForeignKey("flashcard_decks.id", ondelete="CASCADE"), nullable=False, index=True)
    front = Column(Text, nullable=False)
    back = Column(Text, nullable=False)
    difficulty = Column(String(20), default="basic")
    streak = Column(Integer, default=0)
    next_review = Column(DateTime, nullable=True)
    created_at = Column(DateTime, server_default=func.now())

    # ── FSRS spaced-repetition state (Idea 52 — vendored py-fsrs) ──
    # ``next_review`` is the FSRS ``due`` datetime. New cards start with
    # ``state=None`` = Learning/never reviewed (due immediately).
    state = Column(Integer, nullable=True)
    step = Column(Integer, nullable=True)
    stability = Column(Float, nullable=True)
    # FSRS difficulty (1–10). Named ``fsrs_difficulty`` because the existing
    # ``difficulty`` column is the display tier (easy/basic/medium/hard).
    fsrs_difficulty = Column(Float, nullable=True)
    # Successful reviews (legacy repetition counter, display only).
    reps = Column(Integer, default=0)
    # Failed reviews since last success (display only).
    lapses = Column(Integer, default=0)

    deck = relationship("FlashcardDeck", back_populates="cards")

File: frontend/src/services/api/types/flashcard.ts
// Flashcards + leaderboard types.
// Extracted from services/api.ts (F8 split).

// ----- Flashcards types (99-phase plan, Groups 4-5) -----
export interface Flashcard {
  id: number;
  deck_id: number;
  front: string;
  back: string;
  difficulty: string;
  streak: number;
  next_review: string | null;
  created_at: string | null;
}

// FSRS spaced-repetition state (Idea 52 — vendored py-fsrs).
export interface FlashcardSchedule {
  card_id: number;
  next_review: string | null;
  interval_days: number;
  state: number | null;
  stability: number | null;
  fsrs_difficulty: number | null;
  reps: number;
  lapses: number;
  streak: number;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: number;
  name: string;
  username: string | null;
  avatar_class: string | null;
  level: number;
  current_streak: number;
  total_xp: number;
  me: boolean;
}

export interface LeaderboardMe {
  user_id: number;
  total_xp: number;
  rank: number;
  percentile: number;
  total_users: number;
}

export interface LeaderboardResponse {
  items: LeaderboardEntry[];
  me: LeaderboardMe;
}

export interface DueFlashcard {
  id: number;
  deck_id: number;
  front: string;
  back: string;
  difficulty: string;
  next_review: string | null;
  interval_days: number;
  streak: number;
}

export interface FlashcardDeck {
  id: number;
  name: string;
  course_id: number | null;
  created_at: string | null;
  card_count: number;
  cards: Flashcard[];
}
